File: src/jira_dc_mcp/tools/projects.py
"""Project introspection tools."""

from __future__ import annotations

import json
from typing import Any

from ..client import JiraClient


async def list_projects(client: JiraClient) -> str:
    """List all projects with key metadata."""
    projects = await client.list_projects()
    result = [
        {
            "id": p.get("id"),
            "key": p.get("key"),
            "name": p.get("name"),
            "projectTypeKey": p.get("projectTypeKey"),
            "lead": p.get("lead", {}).get("displayName") if p.get("lead") else None,
            "description": (p.get("description") or "")[:200],
        }
        for p in projects
    ]
    return json.dumps(result, indent=2)


async def _get_project_config(client: JiraClient, project_key: str) -> dict[str, Any]:
    """Internal: resolve the full scheme chain for a project.

    Uses project-level endpoints (``/project/{key}/workflowscheme``, etc.)
    which are available on Jira DC 10.x.  Falls back gracefully when an
    endpoint returns 404/405.
    """
    import httpx

    project = await client.get_project(project_key)
    pid = int(project["id"])

    async def _try_get(path: str) -> dict | None:
        try:
            return await client.get(path)
        except httpx.HTTPStatusError:
            return None

    # Resolve workflow scheme via project-level endpoint
    wf_scheme = await _try_get(f"/rest/api/2/project/{project_key}/workflowscheme")

    # Notification, permission, and issue security schemes have project-level endpoints
    notif_scheme = await _try_get(f"/rest/api/2/project/{project_key}/notificationscheme")
    perm_scheme = await _try_get(f"/rest/api/2/project/{project_key}/permissionscheme")
    sec_scheme = await _try_get(f"/rest/api/2/project/{project_key}/issuesecuritylevelscheme")

    return {
        "key": project.get("key"),
        "name": project.get("name"),
        "id": pid,
        "projectTypeKey": project.get("projectTypeKey"),
        "lead": project.get("lead", {}).get("displayName") if project.get("lead") else None,
        "issueTypes": [
            {"id": it["id"], "name": it["name"], "subtask": it.get("subtask", False)}
            for it in project.get("issueTypes", [])
        ],
        "schemes": {
            "workflowScheme": {
                "id": wf_scheme.get("id") if wf_scheme else None,
                "name": wf_scheme.get("name") if wf_scheme else "Default",
                "defaultWorkflow": wf_scheme.get("defaultWorkflow") if wf_scheme else None,
                "issueTypeMappings": wf_scheme.get("issueTypeMappings", {}) if wf_scheme else {},
            },
            "notificationScheme": {
                "id": notif_scheme.get("id") if notif_scheme else None,
                "name": notif_scheme.get("name") if notif_scheme else "Default",
            },
            "permissionScheme": {
                "id": perm_scheme.get("id") if perm_scheme else None,
                "name": perm_scheme.get("name") if perm_scheme else "Default",
            },
            "issueSecurityScheme": {
                "id": sec_scheme.get("id") if sec_scheme else None,
                "name": sec_scheme.get("name") if sec_scheme else None,
            },
        },
    }


async def get_project_config(client: JiraClient, project_key: str) -> str:
    """Full configuration chain for a project."""
    result = await _get_project_config(client, project_key)
    return json.dumps(result, indent=2)


async def get_project_role_members(client: JiraClient, project_key: str) -> str:
    """Get all roles and their members for a project."""
    roles_map = await client.get_project_roles(project_key)
    result = {}
    for role_name, role_url in roles_map.items():
        # Extract role ID from URL
        role_id = int(role_url.rstrip("/").split("/")[-1])
        role_data = await client.get_project_role_actors(project_key, role_id)
        result[role_name] = {
            "id": role_id,
            "actors": [
                {
                    "displayName": a.get("displayName"),
                    "type": a.get("type"),
                    "name": a.get("name"),
                }
                for a in role_data.get("actors", [])
            ],
        }
    return json.dumps(result, indent=2)


async def get_project_components(client: JiraClient, project_key: str) -> str:
    comps = await client.get_project_components(project_key)
    result = [
        {
            "id": c.get("id"),
            "name": c.get("name"),
            "lead": c.get("lead", {}).get("displayName") if c.get("lead") else None,
            "assigneeType": c.get("assigneeType"),
            "description": c.get("description", ""),
        }
        for c in comps
    ]
    return json.dumps(result, indent=2)


async def get_project_versions(client: JiraClient, project_key: str) -> str:
    versions = await client.get_project_versions(project_key)
    result = [
        {
            "id": v.get("id"),
            "name": v.get("name"),
            "released": v.get("released", False),
            "archived": v.get("archived", False),
            "releaseDate": v.get("releaseDate"),
            "description": v.get("description", ""),
        }
        for v in versions
    ]
    return json.dumps(result, indent=2)


async def list_project_categories(client: JiraClient) -> str:
    """List all project categories."""
    categories = await client.list_project_categories()
    result = [
        {
            "id": c.get("id"),
            "name": c.get("name"),
            "description": c.get("description", ""),
        }
        for c in categories
    ]
    return json.dumps(result, indent=2)

File: src/tools/projects.ts
/** Project introspection tools. */

import { z } from "zod";
import { boundedAll, JiraClient } from "../client.js";
import { isHttpStatusError } from "../errors.js";
import { dumps } from "../json.js";
import type { ToolDef } from "./types.js";

/**
 * Resolve the full scheme chain for a project.
 *
 * Uses project-level endpoints (`/project/{key}/workflowscheme`, etc.) which
 * are available on Jira DC 10.x; falls back gracefully on 404/405.
 * Exported for reuse by the analysis tools.
 */
export async function resolveProjectConfig(
  client: JiraClient,
  projectKey: string,
): Promise<Record<string, any>> {
  const project = await client.getProject(projectKey);
  const pid = Number(project.id);

  const tryGet = async (path: string): Promise<any> => {
    try {
      return await client.get(path);
    } catch (e) {
      if (isHttpStatusError(e)) return null;
      throw e;
    }
  };

  const wfScheme = await tryGet(`/rest/api/2/project/${projectKey}/workflowscheme`);
  const notifScheme = await tryGet(`/rest/api/2/project/${projectKey}/notificationscheme`);
  const permScheme = await tryGet(`/rest/api/2/project/${projectKey}/permissionscheme`);
  const secScheme = await tryGet(`/rest/api/2/project/${projectKey}/issuesecuritylevelscheme`);

  // DC has no per-project REST endpoint for these three schemes. The issue type
  // scheme is found by scanning each scheme's associations; the other two come
  // from the ScriptRunner endpoints, which list each scheme's projects. A failed
  // lookup is reported in an `error` field rather than failing the whole tool.
  const isProject = (p: any): boolean => String(p?.key) === project.key;

  const issueTypeScheme = await (async (): Promise<Record<string, any>> => {
    try {
      const schemes = await client.listIssueTypeSchemes();
      const associations = await boundedAll(
        schemes.map((s: any) => () => client.getIssueTypeSchemeAssociations(s.id)),
      );
      const i = associations.findIndex((projects) => (projects ?? []).some(isProject));
      // Unassociated projects fall back to the global default issue type scheme.
      return i >= 0
        ? { id: Number(schemes[i].id), name: schemes[i].name }
        : { id: null, name: "Default" };
    } catch (e: any) {
      return { id: null, name: null, error: String(e?.message ?? e) };
    }
  })();

  const issueTypeScreenScheme = await (async (): Promise<Record<string, any>> => {
    try {
      const schemes = await client.listIssueTypeScreenSchemes();
      const s = schemes.find((x: any) => (x.projects ?? []).some(isProject));
      return s
        ? { id: s.id, name: s.name, mappings: s.mappings ?? [] }
        : { id: null, name: null, mappings: [] };
    } catch (e: any) {
      return { id: null, name: null, mappings: [], error: String(e?.message ?? e) };
    }
  })();

  const fieldConfigurationScheme = await (async (): Promise<Record<string, any>> => {
    try {
      const schemes = await client.listFieldConfigurationSchemes();
      const s = schemes.find((x: any) => (x.projects ?? []).some(isProject));
      // No scheme means every issue type uses the system default field configuration.
      return s
        ? { id: s.id, name: s.name, mappings: s.mappings ?? [] }
        : { id: null, name: "System Default Field Configuration", mappings: [] };
    } catch (e: any) {
      return { id: null, name: null, mappings: [], error: String(e?.message ?? e) };
    }
  })();

  return {
    key: project.key,
    name: project.name,
    id: pid,
    projectTypeKey: project.projectTypeKey,
    lead: project.lead ? (project.lead.displayName ?? null) : null,
    issueTypes: (project.issueTypes ?? []).map((it: any) => ({
      id: it.id,
      name: it.name,
      subtask: it.subtask ?? false,
    })),
    schemes: {
      workflowScheme: {
        id: wfScheme ? (wfScheme.id ?? null) : null,
        name: wfScheme ? wfScheme.name : "Default",
        defaultWorkflow: wfScheme ? (wfScheme.defaultWorkflow ?? null) : null,
        issueTypeMappings: wfScheme ? (wfScheme.issueTypeMappings ?? {}) : {},
      },
      issueTypeScheme,
      issueTypeScreenScheme,
      fieldConfigurationScheme,
      notificationScheme: {
        id: notifScheme ? (notifScheme.id ?? null) : null,
        name: notifScheme ? notifScheme.name : "Default",
      },
      permissionScheme: {
        id: permScheme ? (permScheme.id ?? null) : null,
        name: permScheme ? permScheme.name : "Default",
      },
      issueSecurityScheme: {
        id: secScheme ? (secScheme.id ?? null) : null,
        name: secScheme ? secScheme.name : null,
      },
    },
  };
}

export const projectTools: ToolDef[] = [
  {
    name: "list_projects",
    description: "List all Jira projects with key, name, type, lead.",
    inputShape: {},
    async handler({ client }) {
      const projects = await client.listProjects();
      return dumps(
        projects.map((p: any) => ({
          id: p.id,
          key: p.key,
          name: p.name,
          projectTypeKey: p.projectTypeKey,
          lead: p.lead ? (p.lead.displayName ?? null) : null,
          description: (p.description || "").slice(0, 200),
        })),
      );
    },
  },

  {
    name: "get_project_config",
    description:
      "Get the schemes a project uses, plus its basic details (key, name, type, lead) " +
      "and issue types. Schemes: workflow (with issue-type-to-workflow mappings), " +
      "issue type, issue type screen and field configuration (both with their " +
      "issue-type mappings), permission, notification, issue security. The issue type " +
      "screen and field configuration schemes come from ScriptRunner endpoints; a " +
      "scheme that could not be resolved carries an `error` field.",
    inputShape: { project_key: z.string().describe("Jira project key (e.g. 'PROJ')") },
    async handler({ client }, args) {
      return dumps(await resolveProjectConfig(client, args.project_key));
    },
  },

  {
    name: "get_project_role_members",
    description: "Get all roles and their members/actors for a project.",
    inputShape: { project_key: z.string().describe("Jira project key") },
    async handler({ client }, args) {
      const rolesMap: Record<string, string> = await client.getProjectRoles(args.project_key);
      const result: Record<string, unknown> = {};
      for (const [roleName, roleUrl] of Object.entries(rolesMap)) {
        const roleId = Number(roleUrl.replace(/\/+$/, "").split("/").pop());
        const roleData = await client.getProjectRoleActors(args.project_key, roleId);
        result[roleName] = {
          id: roleId,
          actors: (roleData.actors ?? []).map((a: any) => ({
            displayName: a.displayName,
            type: a.type,
            name: a.name,
          })),
        };
      }
      return dumps(result);
    },
  },

  {
    name: "get_project_components",
    description: "Get components for a project with leads and descriptions.",
    inputShape: { project_key: z.string().describe("Jira project key") },
    async handler({ client }, args) {
      const comps = await client.getProjectComponents(args.project_key);
      return dumps(
        comps.map((c: any) => ({
          id: c.id,
          name: c.name,
          lead: c.lead ? (c.lead.displayName ?? null) : null,
          assigneeType: c.assigneeType,
          description: c.description ?? "",
        })),
      );
    },
  },

  {
    name: "get_project_versions",
    description: "Get versions for a project with release status and dates.",
    inputShape: { project_key: z.string().describe("Jira project key") },
    async handler({ client }, args) {
      const versions = await client.getProjectVersions(args.project_key);
      return dumps(
        versions.map((v: any) => ({
          id: v.id,
          name: v.name,
          released: v.released ?? false,
          archived: v.archived ?? false,
          releaseDate: v.releaseDate,
          description: v.description ?? "",
        })),
      );
    },
  },

  {
    name: "list_project_categories",
    description: "List all project categories used to group projects.",
    inputShape: {},
    async handler({ client }) {
      const categories = await client.listProjectCategories();
      return dumps(
        categories.map((c: any) => ({
          id: c.id,
          name: c.name,
          description: c.description ?? "",
        })),
      );
    },
  },
];
